### PR TITLE
feat(rome_js_analyze): `noCommaOperator`

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Closes: #3932

This implements ESlint's rule [no-sequences](https://eslint.org/docs/latest/rules/no-sequences).

In contrast to *ESLint*, we do not support the option `allowInParentheses` (enabled by default in *ESLint* ; disabled in *Rome*).

I am wondering whether we should rename the rule to `noComaOperator`?

## Test Plan

Unit and doc tests included (copied from *ESLint*).

### CODE REVIEW

**Comment by xunilrj** [comment]
Nice! Thanks.
Given that our node is named `JsSequenceExpression` I would keep the `noSequences`. Although I am not sure about the plural now.

File: crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+impl Rule for NoSequences {
+    type Query = Ast<JsSequenceExpression>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let seq = ctx.query();
+        if let Some(for_stmt) = seq.parent::<JsForStatement>() {
+            let is_in_init = for_stmt
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs:57]: I am torn between doing this or looking into the slot you are inside the `for`.

This should be a comparison between pointers and offsets, but I am not entirely sure how optimized the compiler is going here.

File: crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                .initializer()
+                .filter(|for_init| for_init.syntax() == seq.syntax())
+                .is_some();
+            let is_in_up = for_stmt
+                .update()
+                .filter(|for_up| for_up.syntax() == seq.syntax())
+                .is_some();
+            if is_in_init || is_in_up {
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs:65]: How about moving 

```rust
if is_in_init { return None }
```

to above 

````rust
let is_in_up = ...
````

and avoid doing a expensive operation if is ok already?

File: crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                // Allow coma operator in initializer and update parts of a `for`
+                None
+            } else {
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs:68]: Do you need these two `else`?
You can have a  `Some(())` at the end of the function, right?


````rust
if let Some(for_stmt) = seq.parent::<JsForStatement>() {
    let is_in_init = ...
    if is_in_init { return None }

    let is_in_update = ...
    if is_in_update { return None }
}

Some(())

File: crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs (hunk 1 above)
Comment (Conaclos) [reply crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs]: Is there a practical difference in terms of performance? Internally `initializer` and `update` relies on `slots`? `syntax` is a no-op? and `filter` compiles to a conditional?

Comment (xunilrj) [reply crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs]: Good question. I have never measure it to known what happens in the real world. Let us keep this way.

**Review by leops — COMMENTED** [review]
I personally think we should rename the rule to `noCommaOperator` since "sequence" is used internally as the name of the AST node but may not be clear to the end-user (this can be seen in the documentation for the ESLint rule using the name "comma operator" more prominently than the term "sequence")

File: crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        let seq = ctx.query();
+        if let Some(for_stmt) = seq.parent::<JsForStatement>() {
+            if for_stmt
+                .initializer()
+                .filter(|for_init| for_init.syntax() == seq.syntax())
+                .is_some()
+                || for_stmt
+                    .update()
+                    .filter(|for_up| for_up.syntax() == seq.syntax())
+                    .is_some()
```
Comment (MichaReiser) [inline crates/rome_js_analyze/src/analyzers/nursery/no_sequences.rs:64]: I think this should work too
```suggestion
            if for_stmt.initializer() == Some(
                .filter(|for_init| for_init.syntax() == seq.syntax())
                .is_some()
                || for_stmt
                    .update()
                    .filter(|for_up| for_up.syntax() == seq.syntax())
                    .is_some()
```

Or you can test if the expression isn't the `test` expression (it then is guaranteed to be the update or init):

```
for_stmt.test().map(AstNode::into_syntax).as_ref() == Some(seq.syntax())
```

**Review by MichaReiser — APPROVED** [review]

**Review by leops — APPROVED** [review]

**Pull request merged**